src/V3Timing.cpp

```diff
@@ -0,0 +1,658 @@
+        if (nodep->user1SetOnce()) return;
+        AstNode* const controlp = factorOutTimingControl(nodep);
+        if (!controlp) return;
+        // Handle the intra assignment timing control
+        FileLine* const flp = nodep->fileline();
+        if (VN_IS(nodep, AssignDly)) {
+            // If it's an NBA with an intra assignment delay, put it in a fork
+            auto* const forkp = new AstFork{flp, "", nullptr};
+            forkp->joinType(VJoinType::JOIN_NONE);
+            controlp->replaceWith(forkp);
+            forkp->addStmtsp(controlp);
+        }
+        // Insert new vars before the timing control if we're in a function; in a process we can't
+        // do that. These intra-assignment vars will later be passed to forked processes by value.
+        AstNode* const insertBeforep = VN_IS(m_procp, CFunc) ? controlp : nullptr;
+        // Function for replacing values with intermediate variables
+        const auto replaceWithIntermediate = [&](AstNode* const valuep, const std::string& name) {
+            AstVarScope* const newvscp = createTemp(flp, name, valuep->dtypep(), insertBeforep);
+            valuep->replaceWith(new AstVarRef{flp, newvscp, VAccess::READ});
+            controlp->addHereThisAsNext(
+                new AstAssign{flp, new AstVarRef{flp, newvscp, VAccess::WRITE}, valuep});
+        };
+        // Create the intermediate select vars. Note: because 'foreach' proceeds in
+        // pre-order, and we replace indices in selects with variables, we cannot
+        // reach another select under the index position. This is exactly what
+        // we want as only the top level selects are LValues. As an example,
+        // this transforms 'x[a[i]][b[j]] = y'
+        // into 't1 = a[i]; t0 = b[j]; x[t1][t0] = y'.
+        nodep->lhsp()->foreach<AstSel>([&](AstSel* selp) {
+            if (VN_IS(selp->lsbp(), Const)) return;
+            replaceWithIntermediate(selp->lsbp(), m_intraLsbNames.get(nodep));
+            // widthp should be const
+        });
+        nodep->lhsp()->foreach<AstNodeSel>([&](AstNodeSel* selp) {
+            if (VN_IS(selp->bitp(), Const)) return;
+            replaceWithIntermediate(selp->bitp(), m_intraIndexNames.get(nodep));
+        });
+        // Replace the RHS with an intermediate value var
+        replaceWithIntermediate(nodep->rhsp(), m_intraValueNames.get(nodep));
+    }
+    virtual void visit(AstAssignW* nodep) override {
+        iterateChildren(nodep);
+        if (!nodep->timingControlp()) return;
+        // This assignment will be converted to an always. In some cases this may generate an
+        // UNOPTFLAT, e.g.: assign #1 clk = ~clk; We Create a temp var for the LHS of this assign,
+        // to disable the UNOPTFLAT warning for it
+        // TODO: Find a way to do this without introducing this var. Perhaps make V3SchedAcyclic
+        // recognize awaits and prevent it from treating this kind of logic as cyclic
```

Comment:
@kbieganski Please file an issue to track this based on our discussion. (also please file any issues that you think should be tackled at some later point)